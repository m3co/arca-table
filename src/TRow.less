
table {
  border-collapse: collapse;
  tbody {
    .dirty {
      border: 1px dashed red;
    }
  }

}
