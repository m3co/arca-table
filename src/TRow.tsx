import * as React from 'react';

import { TCell } from './TCell';

import { Field, Row, Info } from 'arca-redux';

import { SearchSocket } from './Search-socket';

import './TRow.less';

interface Props {
  Info: Info;
  Row: Row;
  onEdit?: (Row: Row, column?: keyof Row, Field?: Field) => void;
  onRemove?: (Row: Row) => void;
  dirty?: boolean;
  socket?: SearchSocket;
}

interface State {
  dirty: boolean;
  pk?: string;
}

function getPK(Row: Row, Info: Info): string {
  const PKs: string[] = Info.Fields
    .filter((field): boolean => field.Primary)
    .map((field: Field): string => field.Name);
  const columns = Object.keys(Row) as (keyof typeof Row)[];
  return PKs.map((pk: string): string => {
    const column = columns.find((column): boolean => column === pk);
    return column ? Row[column].toString() : '';
  }).join();
}

export class TRow
  extends React.Component<Props, State>
{
  public state: State = {
    dirty: false,
  }

  public constructor(props: Props) {
    super(props);
    this.state.dirty = !!this.props.dirty;
  }

  public componentDidUpdate(): void {
    const { Row, Info } = this.props;
    const { pk, dirty } = this.state;
    if (pk && dirty) {
      if (pk !== getPK(Row, Info)) {
        this.setState({ dirty: false });
      }
    }
  }

  private remove = (): void => {
    const { onRemove, Row, Info } = this.props;
    const pk = getPK(Row, Info);

    if (onRemove) {
      this.setState({ dirty: true, pk });
      onRemove(Row);
    }
  }

  public render(): JSX.Element {
    const { Info, Row, onEdit, socket: search } = this.props;
    const { dirty } = this.state;
    const className = dirty ? 'dirty' : undefined;
    return (
      <tr className={className}>
        {
          Info.Fields.map((field, key): JSX.Element =>
            <TCell
              key={key}
              Row={Row}
              Field={field}
              onEdit={onEdit}
              socket={search} />
          )
        }
        <td>
          <button onClick={this.remove}>-</button>
        </td>
      </tr>
    );
  }
}
