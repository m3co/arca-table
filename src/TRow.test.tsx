
import * as React from 'react';
import { mount } from 'enzyme';

import { TRow } from './TRow';
import { FACADSchedules, Info, Field, Row, Fields } from 'arca-redux';

interface MockHanders {
  onEdit?: (Row: Row, column?: keyof Fields, Field?: Field) => void;
  onDelete?: (Row: Row) => void;
}

function prepareMock1(overloadedHandlers?: MockHanders): JSX.Element {
  const row: FACADSchedules["Row"] = {
    ID: 1,
    BuiltInCategory: 'INVALID',
    Name: 'A name',
    PathName: 'A path',
  }

  const handlers: MockHanders = overloadedHandlers || {};

  return (
    <table>
      <tbody>
        <TRow Row={row} Info={Info} onEdit={handlers.onEdit} onRemove={handlers.onDelete} />
      </tbody>
    </table>
  );
}

test('Row renders a normal cell', (): void => {
  const el = prepareMock1();
  const wrapper = mount(el);
  expect(wrapper).toMatchSnapshot();
});

test('Click over delete fires the delete function', (): void => {
  let onDeleteReached = false;
  const onDelete = (): void => {
    onDeleteReached = true;
  }
  const el = prepareMock1({ onDelete });
  const wrapper = mount(el);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('button').simulate('click');
  wrapper.update();
  expect(onDeleteReached).toBe(true);
  expect(wrapper).toMatchSnapshot();
})

const Info: Info = {
  "Actions": {
    "Delete": true,
    "Insert": true,
    "Update": true,
  },
  "Fields": [
    {
      "Combobox": null,
      "Editable": false,
      "Name": "ID",
      "Primary": true,
      "Required": false,
      "Select": null,
      "Type": "Number",
    },
    {
      "Combobox": null,
      "Editable": true,
      "Name": "Name",
      "Primary": false,
      "Required": true,
      "Select": null,
      "Type": "String",
    },
    {
      "Combobox": null,
      "Editable": true,
      "Name": "BuiltInCategory",
      "Primary": false,
      "Required": true,
      "Select": null,
      "Type": "String",
    },
    {
      "Combobox": null,
      "Editable": true,
      "Name": "PathName",
      "Primary": false,
      "Required": true,
      "Select": null,
      "Type": "String",
    },
  ],
}
