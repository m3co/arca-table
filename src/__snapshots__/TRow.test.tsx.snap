// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Click over delete fires the delete function 1`] = `
<table>
  <tbody>
    <TRow
      Info={
        Object {
          "Actions": Object {
            "Delete": true,
            "Insert": true,
            "Update": true,
          },
          "Fields": Array [
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
          ],
        }
      }
      Row={
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 1,
          "Name": "A name",
          "PathName": "A path",
        }
      }
      onRemove={[Function]}
    >
      <tr>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="0"
        >
          <td
            className="primary"
            onClick={[Function]}
          >
            1
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="1"
        >
          <td
            onClick={[Function]}
          >
            A name
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="2"
        >
          <td
            onClick={[Function]}
          >
            INVALID
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="3"
        >
          <td
            onClick={[Function]}
          >
            A path
          </td>
        </TCell>
        <td>
          <button
            onClick={[Function]}
          >
            -
          </button>
        </td>
      </tr>
    </TRow>
  </tbody>
</table>
`;

exports[`Click over delete fires the delete function 2`] = `
<table>
  <tbody>
    <TRow
      Info={
        Object {
          "Actions": Object {
            "Delete": true,
            "Insert": true,
            "Update": true,
          },
          "Fields": Array [
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
          ],
        }
      }
      Row={
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 1,
          "Name": "A name",
          "PathName": "A path",
        }
      }
      onRemove={[Function]}
    >
      <tr
        className="dirty"
      >
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="0"
        >
          <td
            className="primary"
            onClick={[Function]}
          >
            1
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="1"
        >
          <td
            onClick={[Function]}
          >
            A name
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="2"
        >
          <td
            onClick={[Function]}
          >
            INVALID
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="3"
        >
          <td
            onClick={[Function]}
          >
            A path
          </td>
        </TCell>
        <td>
          <button
            onClick={[Function]}
          >
            -
          </button>
        </td>
      </tr>
    </TRow>
  </tbody>
</table>
`;

exports[`Row renders a normal cell 1`] = `
<table>
  <tbody>
    <TRow
      Info={
        Object {
          "Actions": Object {
            "Delete": true,
            "Insert": true,
            "Update": true,
          },
          "Fields": Array [
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            },
          ],
        }
      }
      Row={
        Object {
          "BuiltInCategory": "INVALID",
          "ID": 1,
          "Name": "A name",
          "PathName": "A path",
        }
      }
    >
      <tr>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": false,
              "Name": "ID",
              "Primary": true,
              "Required": false,
              "Select": null,
              "Type": "Number",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="0"
        >
          <td
            className="primary"
            onClick={[Function]}
          >
            1
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "Name",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="1"
        >
          <td
            onClick={[Function]}
          >
            A name
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "BuiltInCategory",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="2"
        >
          <td
            onClick={[Function]}
          >
            INVALID
          </td>
        </TCell>
        <TCell
          Field={
            Object {
              "Combobox": null,
              "Editable": true,
              "Name": "PathName",
              "Primary": false,
              "Required": true,
              "Select": null,
              "Type": "String",
            }
          }
          Row={
            Object {
              "BuiltInCategory": "INVALID",
              "ID": 1,
              "Name": "A name",
              "PathName": "A path",
            }
          }
          key="3"
        >
          <td
            onClick={[Function]}
          >
            A path
          </td>
        </TCell>
        <td>
          <button
            onClick={[Function]}
          >
            -
          </button>
        </td>
      </tr>
    </TRow>
  </tbody>
</table>
`;
